### PR TITLE
Add has successors/predecessors methods (on digraph)

### PR DESCRIPTION
**Pull request by harlowja**

### CODE REVIEW

File: networkx/classes/digraph.py [hunk 1]
```diff
@@ -732,6 +732,13 @@ def has_successor(self, u, v):
         """
         return (u in self.succ and v in self.succ[u])
 
+    def has_successors(self, u, v):
```
Comment (jfinkels) [inline networkx/classes/digraph.py]: One too many arguments here.

File: networkx/classes/digraph.py [hunk 2]
```diff
@@ -732,6 +732,13 @@ def has_successor(self, u, v):
...
+        """Return True if node u has any successors, otherwise False."""
```
Comment (jfinkels) [inline networkx/classes/digraph.py]: Needs double backticks around code: ` ``True`` `, ` ``u`` `, etc.

File: networkx/classes/digraph.py [hunk 3]
```diff
@@ -732,6 +732,13 @@ def has_successor(self, u, v):
...
+        try:
+            return bool(self.succ[u])
```
Comment (jfinkels) [inline networkx/classes/digraph.py]: Though it is correct, I find casting a dictionary to a Boolean a little mysterious. It's a very C-like thing to do. `return len(self.succ[u]) == 0` is clearer. I don't feel too too strongly about that though.

File: networkx/classes/digraph.py [hunk 4]
```diff
@@ -732,6 +732,13 @@ def has_successor(self, u, v):
...
+        except KeyError:
+            raise NetworkXError("The node %s is not in the digraph."%(n,))
```
Comment (jfinkels) [inline networkx/classes/digraph.py]: Use a format string (with the correct arguments) `'{}'.format(...)`.

**Comment by jfinkels** [comment]
Makes sense to me.

Needs unit tests and documentation needs to be exposed in `doc/source/reference/classes.digraph.rst`.

**Comment by dschult** [comment]
I'm concerned here about overcrowding the methods on the base classes.
Can you explain why this is needed and explain your thoughts on the interface.
For example, why isn't there a has_neighbors method for undirected graphs?
Also, why is ```if G.has_successors(u)``` better than ```if G.succ[u]```?

**Comment by jfinkels** [comment]
(or `if G[u]`, since the user doesn't know about `G.succ`).

**Comment by MridulS** [comment]
Or just G.successors

File: networkx/classes/digraph.py (hunk 1 above)
Comment (harlowja) [inline networkx/classes/digraph.py]: Oops, thanks, will fix.

File: networkx/classes/digraph.py (hunk 3 above)
Comment (harlowja) [inline networkx/classes/digraph.py]: K, either or.

**Comment by harlowja** [comment]
U guys decide, I thought explicit methods were nice, vs having to access `succ` or `pred` and knowing about the imho internals of the graph classes, I feel `succ` or `pred` are more of internal details, although maybe you guys don't believe that; if that's so then this is needed.

File: networkx/classes/digraph.py (hunk 4 above)
Comment (harlowja) [inline networkx/classes/digraph.py]: Okies.

**Comment by jfinkels** [comment]
> [...] I feel `succ` or `pred` are more of internal details [...]

Yes, they are implementation details, the user should not need to use them. But you could still do the same successors check you are proposing here by `if G[v]:`; this is also why the graph class doesn't need a `has_neighbors()` method. However, the check for predecessors is slightly less simple, maybe `any(v in G[u] for u in G)`. That's why I think this is a valid possibility. Another name for such a method might be `G.is_minimal(v)`.

**Comment by harlowja** [comment]
I'm confused `any(v in G[u] for u in G)` that would mean I'd need to scan all the nodes in a graph to check this, that seems overkill for just checking whether something has a predecessors. This seems like a basic method of a graph IMHO, `if G[v]:` seems ok, didn't know that existed, but having to scan the full graph to check something really simple, seems not optimal.

File: networkx/classes/digraph.py (hunk 2 above)
Comment (harlowja) [inline networkx/classes/digraph.py]: That will blow past the 80 line length, but I'll do a few of them.

**Comment by harlowja** [comment]
Also btw using `successors` or `predecessors` expands an iterator, which again isn't optimal for an existence check, although yes they could be used to do this, but for large graphs, with many edges, why would that be the preferred way?

**Comment by jfinkels** [comment]
> I'm confused `any(v in G[u] for u in G)` that would mean I'd need to scan all the nodes in a graph to check this, that seems overkill for just checking whether something has a predecessors.

Yes, I'm agreeing with you that using `len(G.pred[u]) > 0` is a better implementation and `has_predecessors` has merit enough to exist as a method.

> Also btw using `successors` or `predecessors` expands an iterator, which again isn't optimal for an existence check, although yes they could be used to do this, but for large graphs, with many edges, why would that be the preferred way?

Yep, don't use `predecessors()`.

File: networkx/classes/digraph.py (hunk 2 above)
Comment (jfinkels) [inline networkx/classes/digraph.py]: Then update the docstrings so that they are at most 79 characters.

**Comment by MridulS** [comment]
I'll go ahead and close this. This would expand the base classes and we can do a one line check with `G.pred[node]` and `G.succ[node]` AtlasViews.

``` python
>>> G = nx.DiGraph()
>>> G.add_edge('a', 'b')
>>> if G.succ['a']: # will evaluate to True

>>> if G.pred['a']: # will evaluate to False

```

**Pull request closed without merging**